lib/client/conntest/database.go

```diff
@@ -0,0 +1,394 @@
+type DatabaseConnectionTester struct {
+	cfg          DatabaseConnectionTesterConfig
+	webProxyAddr string
+}
+
+// NewDatabaseConnectionTester returns a new DatabaseConnectionTester
+func NewDatabaseConnectionTester(cfg DatabaseConnectionTesterConfig) (*DatabaseConnectionTester, error) {
+	parsedProxyHostAddr, err := client.ParseProxyHost(cfg.ProxyHostPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &DatabaseConnectionTester{
+		cfg:          cfg,
+		webProxyAddr: parsedProxyHostAddr.WebProxyAddr,
+	}, nil
+}
+
+// TestConnection tests the access to a database using:
+// - auth Client using the User access
+// - the resource name
+// - database user and database name to connect to
+//
+// A new ConnectionDiagnostic is created and used to store the traces as it goes through the checkpoints
+// To connect to the Database, we will create a cert-key pair and setup a Database client back to Teleport Proxy.
+// The following checkpoints are reported:
+// - database server for the requested database exists / the user's roles can access it
+// - the user can use the requested database user and database name (per their roles)
+// - the database is acessible and accepting connections from the database server
+// - the database has the database user and database name that was requested
+func (s *DatabaseConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
+	if req.ResourceKind != types.KindDatabase {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindDatabase, req.ResourceKind)
+	}
+
+	if req.DatabaseUser == "" {
+		return nil, trace.BadParameter("missing required parameter Database User")
+	}
+
+	if req.DatabaseName == "" {
+		return nil, trace.BadParameter("missing required parameter Database Name")
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(
+		connectionDiagnosticID,
+		map[string]string{},
+		types.ConnectionDiagnosticSpecV1{
+			// We start with a failed state so that we don't need to set it to each return statement once an error is returned.
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Lookup the Database Server that's proxying the requested Database.
+	listResourcesResponse, err := s.cfg.UserClient.ListResources(ctx, proto.ListResourcesRequest{
+		PredicateExpression: fmt.Sprintf(`name == "%s"`, req.ResourceName),
+		ResourceType:        types.KindDatabaseServer,
+		Limit:               1,
```

Comment:
I think this issue was discussed in another PR recently :sweat_smile: 

We'll end up with a new method: `GetDatabaseServer`
It will return a single DatabaseServer from a Database Name.

What do you think?

Reply:
It will be nice if we can use the new function. Or maybe can use a large number for now?
https://github.com/gravitational/teleport/blob/4db05674ebd5acd8586c1e4052d7e7c00318d031/lib/web/resources.go#L336